File: GW3/LogFileHandling/Query/Tokens/Token.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    public abstract class Token
    {
        public string Value { get; protected set; }

        abstract public bool CanBeUsed(QueryParser parser);
        abstract public Token Extract(QueryParser parser);
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/Token.ts
abstract class Token {
    public value: string;
    public abstract proposals: SuggestionInterface[];
    public abstract tokenType: TokenType;
    public abstract canBeUsed(parser: QueryParser): boolean;
    public abstract extract(parser: QueryParser): Token;
    public abstract getProposals(nextToken?: Token, afterNextToken?: Token): Token[];
}